CryptoBinPkg/Driver/Packaging/generate_cryptodriver.py

```diff
@@ -33,6 +34,8 @@ def main():
     options = ParseCommandLineOptions()
     # Read in all the functions from the Crypto and Tls headers
     crypto_functions = read_basecryptlib_and_tlslib(options)
+    h_file_path = os.path.join(options.out_dir, "temp_Crypto.h")

```

Comment:
I get why you're using this file to find the EDKII_CRYPTO_VERSION but we actually use the value from CryptoPkg/Include/Protocol/Crypto.h.  I'd leave this as is but we should add a check here that makes sure the versions are identical so we don't forget to update the Crypto.template.h to match Basecore.

Reply:
Here's my understanding. I may have misunderstood the larger context of your comment, please let me know if that's the case.

This is creating `CryptoPkg/Include/Protocol/Crypto.h` at the same time these files are generated based on the value of `options.h_file`:

https://github.com/microsoft/mu_crypto_release/blob/8303fd27ce1494c3612d564bd5febcd0d8512d75/CryptoBinPkg/Driver/Packaging/generate_cryptodriver.py#L33

https://github.com/microsoft/mu_crypto_release/blob/8303fd27ce1494c3612d564bd5febcd0d8512d75/CryptoBinPkg/Driver/Packaging/generate_cryptodriver.py#L43-L48

`options.h_file` is hardcoded to `True`:

https://github.com/microsoft/mu_crypto_release/blob/8303fd27ce1494c3612d564bd5febcd0d8512d75/CryptoBinPkg/Driver/Packaging/generate_cryptodriver.py#L92-L98

And if `options.copy` is set. Which is based on the value of the `--no-copy` parameter which sets the default value of `options.copy` to `True`:

https://github.com/microsoft/mu_crypto_release/blob/8303fd27ce1494c3612d564bd5febcd0d8512d75/CryptoBinPkg/Driver/Packaging/generate_cryptodriver.py#L83-L84

The true (original) value is in `CryptoBinPkg/DriverPackaging/Crypto.template.h`. Since `options.h_file` is hardcoded to `True`, `get_crypto_h()` will always be called which will generate `CryptoBinPkg/Driver/Bin/temp_Crypto.h` from `CryptoBinPkg/DriverPackaging/Crypto.template.h` before `generate_platform_files()` is called.

I look at this as we always use the value that ultimately originates from `Crypto.template.h`. By default, that will end up in `CryptoPkg/Include/Protocol/Crypto.h`. If someone changes that behavior with the `--no-copy` parameter (to not reflect the value in Crypto protocol header), that's on them, but we still say our binaries support what is in the template.

Reply:
Your understanding of the flow from the generate_cryptodriver.py is correct but issues arise in actual practice.  

First, people have typically made changes in MU_BASECORE and then reference their MU_BASECORE branch for generating the crypto binary files.  This leads to some minor changes not being reflected between the two locations (most notably the version variable). 

Second, the script generates every file again but people pick and choose what they include in their update PR.  This can also lead to a version mismatch if we're referencing these temporary files.

A sanity check to confirm they're the same would be nice just so we don't mistakenly publish a version that doesn't actually represent the crypto we're using.  For example, until the other day this would have reported using version 17 when we were definitely using version 18.

Reply:
This is touching on complexities with the circular dependency of the protocol header with the ext_dep in CryptoPkg not directly related to this PR but still something to tighten up.

I will add another change that checks the `CryptoPkg/Include/Protocol/Crypto.h` file against the version in the generated file to basically say, "When you make the ext_dep update with this change you need to update the `Crypto.h` version as well.

I'll do that in another PR though since it's a pre-existing problem.

Reply:
@kenlautner, is that okay with you?

Reply:
That sounds good to me